### PR TITLE
Make Camera::uri required

### PR DESCRIPTION
**Pull request by DmitrySamoylov**
Previously `uri` was used to point to
- file for local cameras
- RTSP stream for IP cameras

and there's inconsistency here: for local cameras we can use `uri` to access camera itself but for IP cams we cannot. So, now `uri` is used to access camera regardless of its type and therefore `uri` is required now.

### CODE REVIEW

File: events/src/camera.rs [hunk 1]
```diff
@@ -1,15 +1,19 @@
...
 /// `Camera` type represents different camera types: IP, USB, CSI cameras
-#[derive(Serialize, Deserialize, Default, Debug, Clone)]
+#[derive(Serialize, Deserialize, Debug, Clone)]
 pub struct Camera {
-    /// ONVIF address if camera is ONVIF-compliant
-    pub onvif_addr: Option<String>,
+    /// URI to access the camera.
+    ///
+    /// - USB or CSI camera: local file name (for example `file:///dev/video0`).
+    /// - IP camera: ONVIF device management service URL (for example `http://192.168.1.2/device`);
```
Comment (zeenix) [inline events/src/camera.rs:15]: Kudos for add docs with examples!

Comment (zeenix) [reply events/src/camera.rs:15]: The commit log says "Previously `uri` was used to point to" but that has remained unchanged so it's confusing. Also,:

> and there's inconsistency here: for local cameras we
> can use `uri` to access camera itself but for IP cams
> we cannot.

I was also very confused by this but then I realized you are talking about the limitation of the API here and not the the cameras themselves. So I'd reword this.

File: events/src/camera.rs [hunk 2]
```diff
@@ -1,15 +1,19 @@
...
+    pub uri: Url,
```
Comment (zeenix) [inline events/src/camera.rs:16]: You're also implicitly changing the type. Ideally that would be a separate change but you should mention it at least in the log.

File: events/src/camera.rs [hunk 3]
```diff
@@ -31,10 +35,8 @@ pub struct Camera {
     /// Physical interface: `usb`, `csi`, `ethernet`
     pub interface: Option<String>,
 
-    /// URI to access video:
-    /// - local file name for `local` cameras (for example `/dev/video0`)
-    /// - RTSP URI for `remote` cameras (for example `rtsp://192.168.1.2/my_stream`)
-    pub uri: Option<String>,
+    /// RTSP stream address (`remote` cameras only).
+    pub rtsp_uri: Option<Url>,
```
Comment (zeenix) [inline events/src/camera.rs:39]: This seems like a hack to have rtsp_uri even if the camera will not have RTSP URI. Why not have separate types for diff types of cameras and have a trait for common API? The `uri` accessor methods in the trait should treat it as optional and the USB camera type can keep it mandatory.

Comment (DmitrySamoylov) [reply events/src/camera.rs:39]: It is just the first PR of the refactoring, I think there'll be enum with all camera types.

File: events/src/camera.rs (hunk 2 above)
Comment (DmitrySamoylov) [reply events/src/camera.rs:16]: Yeah, will split into a separate commit when rebasing.

File: events/src/camera.rs (hunk 1 above)
Comment (DmitrySamoylov) [reply events/src/camera.rs:15]: I wouldn't say it remained unchanged because I changed the comment :)

> limitation of the API

Yep, it's about API of course, will reword.

File: events/src/camera.rs (hunk 3 above)
Comment (zeenix) [reply events/src/camera.rs:39]: cool. Sorry for being pedantic but how does this change lead to/help with the goal of separate types?

Comment (DmitrySamoylov) [reply events/src/camera.rs:39]: Hm, let me think. This change is not about splitting `Camera` into separate types, it's about not loosing RTSP URI in cases when it's needed. There's a bit of [code](https://github.com/lumeohq/lumeod/pull/39/files#diff-c8d72bf2092bec87ffa983a761c12507R54) in lumeod that is used for filling this exact field, just don't want to touch it for now (aside from renaming `uri` to `rtsp_uri`).

**Comment by DmitrySamoylov** [comment]
Since all change requests could be fixed only through rebasing I made one :)

Please have a look again.

File: events/src/camera.rs (hunk 3 above)
Comment (zeenix) [reply events/src/camera.rs:41]: > This change is not about splitting Camera into separate types

Yeah, I understand. What I was saying was that this seems very much like a hack that becomes redundant with the proper fix of splitting the type. So unless there is some urgency here, I'd rather we already implement the proper fix instead.

Comment (DmitrySamoylov) [reply events/src/camera.rs:41]: Well, let's see how it looks when I split types then. So, I close this PR.

Comment (zeenix) [reply events/src/camera.rs:41]: Thanks and sorry for being annoying. :)
